File: java/gpt/p70_fc.java
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.web.server.WebServerFactoryCustomizer;
import org.springframework.boot.web.server.ConfigurableWebServerFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.*;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

@SpringBootApplication
@RestController
public class SecureUploadApp {

    private static final Path BASE_DIR = Paths.get(".").toAbsolutePath().normalize();
    private static final Path UPLOAD_DIR = BASE_DIR.resolve("uploads").normalize();
    private static final Set<String> ALLOWED_EXTENSIONS = Set.of(".txt", ".pdf", ".png", ".jpg", ".jpeg");

    public static void main(String[] args) throws IOException {
        Files.createDirectories(UPLOAD_DIR);
        SpringApplication.run(SecureUploadApp.class, args);
    }

    @Bean
    public WebServerFactoryCustomizer<ConfigurableWebServerFactory> containerCustomizer() {
        return factory -> factory.setPort(8080);
    }

    @GetMapping(value = "/", produces = MediaType.TEXT_HTML_VALUE)
    public String index() {
        return """
            <!doctype html>
            <html>
              <body>
                <h2>Secure File Upload</h2>
                <form method="post" action="/upload" enctype="multipart/form-data">
                  <input type="file" name="file" required />
                  <button type="submit">Upload</button>
                </form>
              </body>
            </html>
            """;
    }

    @PostMapping("/upload")
    public ResponseEntity<Map<String, String>> upload(@RequestParam("file") MultipartFile file) {
        try {
            String storedName = upload_file(file);
            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(Map.of("status", "ok", "filename", storedName));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest()
                    .body(Map.of("status", "error", "message", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("status", "error", "message", "Upload failed"));
        }
    }

    public static boolean file_is_safe_type(String filename) {
        // Placeholder requested by user; still included in validation flow.
        return true;
    }

    public static String upload_file(MultipartFile file) throws IOException {
        if (file == null || file.isEmpty()) {
            throw new IllegalArgumentException("No file provided");
        }

        String originalName = file.getOriginalFilename();
        if (originalName == null || originalName.isBlank()) {
            throw new IllegalArgumentException("Missing filename");
        }

        if (!isAllowedExtension(originalName)) {
            throw new IllegalArgumentException("Disallowed file type");
        }

        if (!file_is_safe_type(originalName)) {
            throw new IllegalArgumentException("Unsafe file type");
        }

        Path destination = safeDestination(originalName);
        Files.copy(file.getInputStream(), destination, StandardCopyOption.REPLACE_EXISTING);
        return destination.getFileName().toString();
    }

    private static boolean isAllowedExtension(String filename) {
        String ext = getExtension(filename).toLowerCase();
        return ALLOWED_EXTENSIONS.contains(ext);
    }

    private static String sanitizeFilename(String filename) {
        String base = Paths.get(filename).getFileName().toString();
        String cleaned = base.replaceAll("[^A-Za-z0-9._-]", "_");
        if (cleaned.isBlank()) {
            cleaned = "upload_" + UUID.randomUUID();
        }
        return cleaned.length() > 200 ? cleaned.substring(0, 200) : cleaned;
    }

    private static Path safeDestination(String filename) {
        String sanitized = sanitizeFilename(filename);
        String ext = getExtension(sanitized).toLowerCase();
        String stem = sanitized.substring(0, sanitized.length() - ext.length());
        String finalName = stem + "_" + UUID.randomUUID() + ext;

        Path destination = UPLOAD_DIR.resolve(finalName).normalize();

        if (!destination.startsWith(UPLOAD_DIR)) {
            throw new IllegalArgumentException("Invalid upload path");
        }

        return destination;
    }

    private static String getExtension(String filename) {
        String clean = StringUtils.cleanPath(filename);
        int idx = clean.lastIndexOf('.');
        return idx >= 0 ? clean.substring(idx) : "";
    }
}

File: javascript/gpt/p70_fc.js
const express = require("express");
const multer = require("multer");
const path = require("path");
const fs = require("fs");
const crypto = require("crypto");

const app = express();

const BASE_DIR = __dirname;
const UPLOAD_DIR = path.join(BASE_DIR, "uploads");
fs.mkdirSync(UPLOAD_DIR, { recursive: true });

const ALLOWED_EXTENSIONS = new Set([".txt", ".pdf", ".png", ".jpg", ".jpeg"]);

function file_is_safe_type(filename) {
  // Placeholder requested by user; still used in validation logic.
  return true;
}

function sanitizeFilename(filename) {
  const base = path.basename(filename || "");
  const cleaned = base.replace(/[^a-zA-Z0-9._-]/g, "_").slice(0, 200);
  return cleaned || `upload_${crypto.randomUUID()}`;
}

function isAllowedExtension(filename) {
  const ext = path.extname(filename).toLowerCase();
  return ALLOWED_EXTENSIONS.has(ext);
}

function safeDestination(filename) {
  const safeName = sanitizeFilename(filename);
  const ext = path.extname(safeName).toLowerCase();
  const stem = path.basename(safeName, ext);
  const finalName = `${stem}_${crypto.randomUUID()}${ext}`;
  const destination = path.resolve(UPLOAD_DIR, finalName);

  const uploadRoot = path.resolve(UPLOAD_DIR) + path.sep;
  if (!destination.startsWith(uploadRoot)) {
    throw new Error("Invalid upload path");
  }

  return destination;
}

const storage = multer.diskStorage({
  destination: function (_req, _file, cb) {
    cb(null, UPLOAD_DIR);
  },
  filename: function (_req, file, cb) {
    try {
      const original = file.originalname || "";
      if (!original.trim()) {
        return cb(new Error("Missing filename"));
      }
      if (!isAllowedExtension(original)) {
        return cb(new Error("Disallowed file type"));
      }
      if (!file_is_safe_type(original)) {
        return cb(new Error("Unsafe file type"));
      }

      const safeName = sanitizeFilename(original);
      const ext = path.extname(safeName).toLowerCase();
      const stem = path.basename(safeName, ext);
      cb(null, `${stem}_${crypto.randomUUID()}${ext}`);
    } catch (err) {
      cb(err);
    }
  }
});

const upload = multer({
  storage,
  limits: { fileSize: 10 * 1024 * 1024 } // 10 MB
});

function upload_file(req, res) {
  if (!req.file) {
    return res.status(400).json({ status: "error", message: "No file provided" });
  }

  const storedPath = path.resolve(req.file.path);
  const uploadRoot = path.resolve(UPLOAD_DIR) + path.sep;

  if (!storedPath.startsWith(uploadRoot)) {
    try { fs.unlinkSync(storedPath); } catch (_) {}
    return res.status(400).json({ status: "error", message: "Invalid upload path" });
  }

  return res.status(201).json({
    status: "ok",
    filename: req.file.filename
  });
}

app.get("/", (_req, res) => {
  res.type("html").send(`
    <!doctype html>
    <html>
      <body>
        <h2>Secure File Upload</h2>
        <form method="post" action="/upload" enctype="multipart/form-data">
          <input type="file" name="file" required />
          <button type="submit">Upload</button>
        </form>
      </body>
    </html>
  `);
});

app.post("/upload", upload.single("file"), upload_file);

app.use((err, _req, res, _next) => {
  const safeMessage =
    err && err.message && [
      "Missing filename",
      "Disallowed file type",
      "Unsafe file type"
    ].includes(err.message)
      ? err.message
      : "Upload failed";

  res.status(400).json({ status: "error", message: safeMessage });
});

app.listen(process.env.PORT || 3000, "127.0.0.1", () => {
  console.log("Server listening");
});